File: typescript/food-ordering/app/src/restate-app/order_status_service.ts
import { OrderStatus, Status } from "./types/types";
import * as restate from "@restatedev/restate-sdk";

const ORDER_STATUS = "order_status";

/** Gets called by the webUI frontend to display the status of an order. */
const get = async (ctx: restate.RpcContext, _orderId: string) =>
  ctx.get<OrderStatus>(ORDER_STATUS);

const setStatus = async (
  ctx: restate.RpcContext,
  orderId: string,
  status: Status
) => {
  console.info(`[${orderId}] Order is ${status}`);
  const currentStatus = await ctx.get<OrderStatus>(ORDER_STATUS);
  ctx.set(ORDER_STATUS, { ...currentStatus, ...{ status } });
};

const setETA = async (
  ctx: restate.RpcContext,
  orderId: string,
  eta: number
) => {
  const currentStatus = await ctx.get<OrderStatus>(ORDER_STATUS);
  ctx.set(ORDER_STATUS, {
    eta: eta,
    status: eta === 0 ? Status.DELIVERED : currentStatus?.status,
  });
};

const handleDriverUpdate = async (
  ctx: restate.RpcContext,
  event: restate.Event
) => {
  const eta = +event.body().toString();
  const currentStatus = await ctx.get<OrderStatus>(ORDER_STATUS);
  ctx.set(ORDER_STATUS, {
    eta: eta,
    status: eta === 0 ? Status.DELIVERED : currentStatus?.status,
  });
};

export const router = restate.keyedRouter({
  get,
  setStatus,
  setETA,
  eventHandler: restate.keyedEventHandler(handleDriverUpdate),
});
export type orderStatusApi = typeof router;
export const service: restate.ServiceApi<orderStatusApi> = {
  path: "orderstatus",
};

File: java/food-ordering/app/restate-app/src/main/java/examples/order/OrderStatusService.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package examples.order;

import static examples.order.generated.OrderProto.*;

import dev.restate.sdk.KeyedContext;
import dev.restate.sdk.common.CoreSerdes;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;
import examples.order.generated.OrderStatusServiceRestate;
import examples.order.types.StatusEnum;

public class OrderStatusService
    extends OrderStatusServiceRestate.OrderStatusServiceRestateImplBase {
  private static final StateKey<Long> ORDER_ETA = StateKey.of("order-eta", CoreSerdes.JSON_LONG);

  /** Gets called by the webUI frontend to display the status of an order. */
  @Override
  public OrderStatus get(KeyedContext ctx, OrderId request) throws TerminalException {
    var status = new OrderWorkflowRestateClient(ctx, request.getOrderId())
            .getState(OrderWorkflow.STATUS)
            .await()
            .orElse(StatusEnum.NEW);
    var eta = ctx.get(ORDER_ETA).orElse(-1L);
    return OrderStatus.newBuilder()
        .setOrderId(request.getOrderId())
        .setStatus(Status.forNumber(status.getValue()))
        .setEta(eta)
        .build();
  }

  @Override
  public void setETA(KeyedContext ctx, NewOrderETA request) throws TerminalException {
    ctx.set(ORDER_ETA, request.getEta());
  }
}
